Eofunc no lapack

maboualidev opened the pull request:
Implements eofunc from NCL.
Does not depends on the locally provided LAPACK and BLAS library.

Unit Test Results:
```
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: test_linint2
PASS: test_eofunc_01
PASS: test_eofunc_02
PASS: test_eofunc_03
PASS: test_eofunc_04
PASS: test_eofunc_05
============================================================================
Testsuite summary for NComp 0.1
============================================================================
# TOTAL: 6
# PASS:  6
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
make[1]: Nothing to be done for `check-am'.
```

clyne commented on src/util.cpp:
```diff
@@ -114,39 +117,70 @@ void convert_to(void *in_arr, size_t in_arr_size, size_t in_arr_offset,
+// So nothing is copied. Hence, you need to make sure that in_arr is not scoped
+// out. However, if the arr_type and the intended are not the same, then a brand
+// new memory locaiton is assigned and the in_arr is copied and converted to it.
+// Hence, if the variable is no longer needed you would need to delete it manually.
```
@abb58 Can we make use of smart pointers here to make the memory ownership/management more robust?

maboualidev replied:
This function, i.e. ```convert_to_with_copy_avoiding(...)``` works on the input data.
(1) If the input data is of type ```double```, we are going to use the input data, which its memory is managed by NumPy from the Python side. So, we should not release the memory in our C/C++ ncomp level.

(2) if the input data is not of type double, then we make a new memory allocation suitable for a double array of the same type and we copy the input data to that new memory. We release the memory at the end of our C/C++ code.

Currently in the code I have:

```C
double * dx = convert_to_with_copy_avoiding<double>(...)
```

and 

at the end of ```eofunc()``` right before existing I have:

```C
if(x_in->type != NCOMP_DOUBLE) delete[] dx;
```

clyne replied:
Any thoughts on this @abb58 ?

abb58 replied:
If the memory is allocated outside a given function (i.e. inside utilities) as it is being done now, there are more restrictions on the scope of the variable. Hence the above described procedure is justified for using manual memory management.

maboualidev replied:
In some other part of the code I use a pattern like this:

```C++
std::unique_ptr<double[]> eval (allocateAndInit<double>(size, initValue));
```

where inside ``allocateAndInit`` a chunk of memory is requested using ``new``. I use this pattern for the working arrays that they need to be cleaned up after we are done with the calculation.

Just in case, if it turns out that that working array needs to be returned to the user, for example as attribute based on the options that the user provided, I call ``.release()`` when I am creating the attribute, here is an example:

```C++
tmp_attr_out.push_back(create_ncomp_single_attribute((char *) "eval_transpose", eval.release(), NCOMP_DOUBLE, 1, dsizes_evec.get()));
```

I could do the same thing for ``dx``, i.e.:

```C++
std::unique_ptr<double[]> dx (convert_to_with_copy_avoiding<double>(...));
```

and at the end do the following:

```C++
// Note that if x_in is already a double, dx is using the same memory
// as in the input; hence, copy avoidance. Therefore, we have to 
// prevent unique_ptr from deleting the object
if(x_in->type != NCOMP_DOUBLE) dx.release();
```

But I don't find this any easier or better than doing:

```C++
double * dx = convert_to_with_copy_avoiding<double>(...)
```
followed by:

```C++
if(x_in->type != NCOMP_DOUBLE) delete[] dx;
```

I prefer to use a simpler data type/structure if I don't get any advantage by using the more complicated one.

clyne replied:
Sounds good. Thanks for the clarification.

clyne commented on src/util.cpp:
```diff
@@ -609,6 +624,218 @@ void _ncomp_coerce(void *from_ptr, int from_type, void *from_missing,
+
+void collectAttributeList(std::vector<ncomp_single_attribute*> attrVector, ncomp_attributes * collectedAttributedList) {
+  collectedAttributedList->nAttribute = attrVector.size();
+  collectedAttributedList->attribute_array  = new ncomp_single_attribute*[attrVector.size()];
```
Use smart pointers or c++ containers, or does this have to play well with C?

maboualidev replied:
This needs to play well with C. Anything that goes through ```collectAttributeList()``` essentially needs to be passed to Python level; Hence, there is no need to call ```delete``` or ```free```.

Essentially, we are using a ```std::vector``` to collect the attributes. The reason that I use ```std::vector``` is that I don't know how many attributes I am going to attach to the output. That depends on the options and configuration that the user is requesting on the input.

On the output, all these different attributes are collected and returned in a Cython/C defined structured which is essentially a pointer to pointer (no copy).

clyne replied:
Understood. Thanks.

maboualidev commented:
@khallock @abb58 @erogluorhan Could you please review this PR and let me know about your decision: approval, request changes, or having more question?

I am blocked by this PR. Need to have this finalized before I could proceed with other functions in eofunc family.

clyne commented:
We actually have several outstanding PRs that need to be reviewed in both https://github.com/NCAR/ncomp and https://github.com/NCAR/geocat-comp. So please everyone be sure to check and make sure there are no PRs being held up by your review. Also, be sure to “resolve” any inline comments that you have made: if you initiated a comment on a PR it is the initiator’s responsibility to ultimately mark the comment as resolved.

thanks!
jc


> On Oct 2, 2019, at 8:59 AM, Mohammad Abouali <notifications@github.com> wrote:
> 
> @khallock @abb58 @erogluorhan Could you please review this PR and let me know about your decision: approval, request changes, or having more question?
> 
> I am blocked by this PR. Need to have this finalized before I could proceed with other functions in eofunc family.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

John Clyne
National Center for Atmospheric Research
303.497.1236 (w), 303.809.1922 (c)
clyne@ucar.edu

abb58 commented on src/util.cpp:
```diff
@@ -114,39 +117,70 @@ void convert_to(void *in_arr, size_t in_arr_size, size_t in_arr_offset,
+  // that this copy is happening.
   int i;
+  new_array->shape = (size_t *) malloc(sizeof(size_t)*array_ndim);
   for (i = 0; i < array_ndim; i++)
```
Define loop variables inside

maboualidev replied:
done.

abb58 commented on src/util.cpp:
```diff
@@ -114,39 +117,70 @@ void convert_to(void *in_arr, size_t in_arr_size, size_t in_arr_offset,
+// Copies one array to another. Note: the to array must have proper shape size.
+// Otherwise, this won't work. Suggesting to change the ncomp_array defintion of
+// shape from size_t[1] to size_t *.
+extern "C" void ncomp_array_copy(ncomp_array * from, ncomp_array * to) {
```
Wouldn't this work ? `to = from`.

Also is it safe to do a shallow copy if one has dynamically allocated pointers in the struct ?

maboualidev replied:
Actually I don't think we need this function anymore. I am going to remove it.

maboualidev replied:
Actually we do need it.

So notice that ``to`` and ``from`` are both of type ``ncomp_array *``, so to answer you question: no, ```to = from``` is not the same as:

```C
to->field_1 = from->field_1;
to->field_2 = from->field_2;
...
to->field_n = from->field_n;
```

Particularly in this case that ```to``` is created in one scope, and ```from``` is created in another scope. the ```to = from``` won't work. If you had time and interested and come to my office and I can explain it in details why it needs to be the way it is.

Also, I am not doing the deep copy, because ``ncomp_array`` has data pointer which could be big. Prefer not to duplicate.

maboualidev replied:
I changed it to 

```C
*to = *from;
```

As mentioned in the meeting,

```C
to = from;
```

will not work.

maboualidev commented on src/ncomp/types.h:
```diff
@@ -2,6 +2,12 @@
 #define NCOMP_TYPES_H
 
 #include <stddef.h>
+#include <limits.h>
```
Wasn't aware of this suggestion/requests. It's done anyway.

maboualidev commented:
Hi @abb58,

Please find the following response to your questions:

**1. Fortran files have been left with F77 format and not F90 ?**
**1.Response:** done

**2. Adopt stringent format for CXX, (type) to static_cast<type>, printf to std::cout, etc**
**2.Response:** Done; I changed all casting to use ```static_cast<type>``` and all the ```printf``` to ```std::cout``` as requested.

However, regarding ```std::cout```,  I personally prefer:
```C
printf("%s msg: %f\n", name, in->msg.msg_int);
```
over
 
```C++
std::cout << name << " msg: " << in->msg.msg_int << std::end;
```
I find the C style more readable than the C++.

Anyhow, I changed the ``printf`` to ``std::out`` as you requested.

**3. Wrap all the sanity checks and print statements into DEBUG mode, since printing to screen from CXX is not advised yet.**
**3.Response:** Done. However, instead of defining a ```DEBUG``` variable in one of the header files, I suggest to pass it at the compile time. we could have ```-DDEBUG``` passed when compiling. I made the comment also on the ``rcm2rgrid`` PR.

Also, We do need to do the checks and stop the calculations if the tests doesn't pass. Regardless of the debug or non-debug mode. Therefore, I wrapped only the error message printing in tho the debug mode.

**4. Check the workings of convert_to_with_copy_avoiding aligns with appropriate routines in NCL.**
**4.Response:** I think we already discussed this together and concluded that both routines are returning the same thing.


So, please review the code changes and let me know about your decision.

Regards,

Moe

abb58 commented:
> Hi @abb58,
> 
> Please find the following response to your requested changes:
> 
> **1. Fortran files have been left with F77 format and not F90 ?**
> **1.Response:** done
> 
> **2. Adopt stringent format for CXX, (type) to static_cast, printf to std::cout, etc**
> **2.Response:** Done; I changed all casting to use `static_cast<type>` and all the `printf` to `std::cout` as requested.
> 
> However, regarding `std::cout`, I personally prefer:
> 
> ```c
> printf("%s msg: %f\n", name, in->msg.msg_int);
> ```
> 
> over
> 
> ```c++
> std::cout << name << " msg: " << in->msg.msg_int << std::end;
> ```
> 
> I find the C style more readable than the C++.
> 
> Anyhow, I changed the `printf` to `std::out` as you requested.
> 

> **3. Wrap all the sanity checks and print statements into DEBUG mode, since printing to screen from CXX is not advised yet.**
> **3.Response:** Done. However, instead of defining a `DEBUG` variable in one of the header files, I suggest to pass it at the compile time. we could have `-DDEBUG` passed when compiling. I made the comment also on the `rcm2rgrid` PR.
> 
> Also, We do need to do the checks and stop the calculations if the tests doesn't pass. Regardless of the debug or non-debug mode. Therefore, I wrapped only the error message printing in tho the debug mode.
> 
> > `DEBUG` in the local file was added as a temporary solution to get things running before necessary infrastructure is added to `autotools`. Feel free to add it. 

> **4. Check the workings of convert_to_with_copy_avoiding aligns with appropriate routines in NCL.**
> **4.Response:** I think we already discussed this together and concluded that both routines are returning the same thing.
> 
> > I have already expressed the workflow with respect to the function `coerce_input_X` over chat and email. Get an opinion/consensus with other fellow members. 

> So, please review the code changes and let me know about your decision.
> 
> Regards,
> 
> Moe
